### PR TITLE
Add support for c++/cli

### PR DESCRIPTION
**Pull request by peters**
I cleaned up the revision history and created a new branch for this feature.

History of previous PR is available here: https://github.com/alexhultman/uWebSockets/pull/42

/cc @alexhultman

### CODE REVIEW

File: src/uWS.h [hunk 1]
```diff
@@ -1,11 +1,30 @@
...
+#if HAVE_WIN32_CPP_CLI
+#define USE_WIN32_THREADS 1
+#else
+#define USE_WIN32_THREADS 0
+#endif
+#endif
+
 #include <cstddef>
 #include <functional>
-#include <string>
```
Comment (peters) [inline src/uWS.h]: @alexhultman Unused include.

File: src/uWS.h [hunk 2]
```diff
@@ -1,11 +1,30 @@
 #ifndef UWS_H
 #define UWS_H
 
+#if defined(_WIN32) && defined(__cplusplus_cli)
+#define HAVE_WIN32_CPP_CLI 1
```
Comment (alexhultman) [inline src/uWS.h]: Why aren't you just using #ifdef __cplusplus_cli directly?

File: src/uWS.cpp [hunk 3]
```diff
@@ -1,3 +1,7 @@
+#if HAVE_WIN32_CPP_CLI
+#error "c++/cli does not support WIN32 threading model"
+#endif
```
Comment (alexhultman) [inline src/uWS.cpp]: So you are checking if the .cpp is being compiled with /clr here?

File: src/uWS.h (hunk 2 above)
Comment (peters) [inline src/uWS.h]: No reason in particular, just to imply that this was Windows related. Shall i remove it?

File: src/uWS.cpp (hunk 3 above)
Comment (peters) [inline src/uWS.cpp]: Correct.

File: src/uWS.h [hunk 4]
```diff
@@ -1,11 +1,30 @@
...
+#else
+#ifndef HAVE_WIN32_CPP_CLI 
+#define HAVE_WIN32_CPP_CLI 0
+#endif
+#endif
+
+#ifndef USE_WIN32_THREADS
+#if HAVE_WIN32_CPP_CLI
+#define USE_WIN32_THREADS 1
+#else
```
Comment (alexhultman) [inline src/uWS.h]: USE_WIN32_THREADS is the same as __cplusplus_cli

File: src/uWS.h [hunk 5]
```diff
@@ -14,8 +33,51 @@
...
+#endif
+  
+  MutexImpl(const MutexImpl& other) = delete;
+  MutexImpl& operator=(const MutexImpl& other) = delete;
+
+public:
+  MutexImpl() {
+#if USE_WIN32_THREADS
+    InitializeCriticalSection(&m_mutex);
+#endif
```
Comment (alexhultman) [inline src/uWS.h]: Like I said, from Googling the problem, StackOverflow says that you should use .NET threading when in C++/CLI. This is using the Win32 threading functions.

File: src/uWS.h (hunk 4 above)
Comment (peters) [inline src/uWS.h]: Some people use ancient visual studio versions and in order to use `mutex` you need at least VS 2012. If you don't care about those (I myself certainly don't) then I'll gladly remove it :)

File: src/uWS.h [hunk 6]
```diff
@@ -1,11 +1,30 @@
 #ifndef UWS_H
 #define UWS_H
 
+#if defined(_WIN32) && defined(__cplusplus_cli)
+#define HAVE_WIN32_CPP_CLI 1
+#pragma managed(push, off)
```
Comment (alexhultman) [inline src/uWS.h]: Is this marking the current file as unmanaged?

Comment (peters) [inline src/uWS.h]: Correct.

File: src/uWS.h (hunk 5 above)
Comment (peters) [inline src/uWS.h]: I'll fix it.

**Comment by alexhultman** [comment]
From this petition: https://visualstudio.uservoice.com/forums/121579-visual-studio-2015/suggestions/3814789-support-c-11-concurrency-header-files-in-c-cli

I can navigate to comments from Microsoft about how they have "unblocked" some C++11 STL features in Visual Studio 2015. What this means is really that Microsoft has now finally *implemented* parts of C++11 for their C++/CLI platform. In other words, this is up to Microsoft to fix and this is not just about mutex but also thread, atomic, condition_variable, etc. I cannot accept workarounds based on Microsofts own incompetence with the C++ standard. This code is not set in stone and will change over time - I cannot have workarounds for standard parts of the C++ language based on Microsofts incapacity to follow ISO standards. Microsoft has always been at least 7 years behind everyone else when it comes to C++ and I cannot have them stall my use of standard features. When I start using C++17 it will break in other places and then we have to fix that up.

**Comment by peters** [comment]
That's okay! I'll just maintain a separate branch and try to stay on top of your changes. Thank you for taking the time to review my changes at least :)

**Comment by alexhultman** [comment]
Thanks for the interest

**Pull request closed without merging**